File: src/main/java/net/kronos/rkon/core/Rcon.java
package net.kronos.rkon.core;

import java.io.IOException;
import java.net.Socket;
import java.nio.charset.Charset;
import java.util.Random;

import net.kronos.rkon.core.ex.AuthenticationException;

public class Rcon {
	
	private final Object sync = new Object();
	private final Random rand = new Random();
	
	private int requestId;
	private Socket socket;
	
	private Charset charset;

	/**
	 * Create, connect and authenticate a new Rcon object
	 * 
	 * @param host Rcon server address
	 * @param port Rcon server port
	 * @param password Rcon server password
	 * 
	 * @throws IOException
	 * @throws AuthenticationException
	 */
	public Rcon(String host, int port, byte[] password) throws IOException, AuthenticationException {
		// Default charset is utf8
		this.charset = Charset.forName("UTF-8");
		
		// Connect to host
		this.connect(host, port, password);
	}
	
	/**
	 * Connect to a rcon server
	 * 
	 * @param host Rcon server address
	 * @param port Rcon server port
	 * @param password Rcon server password
	 * 
	 * @throws IOException
	 * @throws AuthenticationException
	 */
	public void connect(String host, int port, byte[] password) throws IOException, AuthenticationException {
		if(host == null || host.trim().isEmpty()) {
			throw new IllegalArgumentException("Host can't be null or empty");
		}
		
		if(port < 1 || port > 65535) {
			throw new IllegalArgumentException("Port is out of range");
		}
		
		// Connect to the rcon server
		synchronized(sync) {
			// New random request id
			this.requestId = rand.nextInt();
			
			// We can't reuse a socket, so we need a new one
			this.socket = new Socket(host, port);
		}
		
		// Send the auth packet
		RconPacket res = this.send(RconPacket.SERVERDATA_AUTH, password);
		
		// Auth failed
		if(res.getRequestId() == -1) {
			throw new AuthenticationException("Password rejected by server");
		}
	}
	
	/**
	 * Disconnect from the current server
	 * 
	 * @throws IOException
	 */
	public void disconnect() throws IOException {
		synchronized(sync) {
			this.socket.close();
		}
	}
	
	/**
	 * Send a command to the server
	 * 
	 * @param payload The command to send
	 * @return The payload of the response
	 * 
	 * @throws IOException
	 */
	public String command(String payload) throws IOException {
		if(payload == null || payload.trim().isEmpty()) {
			throw new IllegalArgumentException("Payload can't be null or empty");
		}
		
		RconPacket response = this.send(RconPacket.SERVERDATA_EXECCOMMAND, payload.getBytes());
		
		return new String(response.getPayload(), this.getCharset());
	}
	
	private RconPacket send(int type, byte[] payload) throws IOException {
		synchronized(sync) {
			return RconPacket.send(this, type, payload);
		}
	}

	public int getRequestId() {
		return requestId;
	}
	
	public Socket getSocket() {
		return socket;
	}
	
	public Charset getCharset() {
		return charset;
	}
	
	public void setCharset(Charset charset) {
		this.charset = charset;
	}

}

File: src/main/kotlin/net/kronos/rkon/core/Rcon.kt
package net.kronos.rkon.core

import io.github.starwishsama.comet.utils.BotUtil
import net.kronos.rkon.core.ex.AuthenticationException
import java.io.IOException
import java.net.Socket
import java.nio.charset.Charset
import java.util.*

class Rcon(host: String, port: Int, password: ByteArray?) {
    private val sync = Any()
    private val rand = Random()
    var requestId = 0
        private set
    var socket: Socket? = null
        private set
    var charset: Charset = Charset.forName("UTF-8")

    /**
     * Connect to a rcon server
     *
     * @param host Rcon server address
     * @param port Rcon server port
     * @param password Rcon server password
     *
     * @throws IOException
     * @throws AuthenticationException Thrown if server rejects password
     */
    @Throws(IOException::class, AuthenticationException::class)
    fun connect(host: String, port: Int, password: ByteArray?) {
        require(host.trim { it <= ' ' }.isNotEmpty()) { "Host can't be null or empty" }
        require(!(port < 1 || port > 65535)) { "Port is out of range" }

        // Connect to the rcon server
        synchronized(sync) {

            // New random request id
            requestId = rand.nextInt()

            // We can't reuse a socket, so we need a new one
            socket = Socket(host, port)
        }

        // Send the auth packet
        // Anonymous sendAuth(Rcon rcon, byte[] password) for one-time use
        var response: RconPacket?
        synchronized(sync) { response = password?.let { RconPacket.sendAuth(this, it) } }


        // Check auth status
        if (response?.requestId != requestId) {
            if (response?.requestId == -1) {
                // Auth failed
                throw AuthenticationException("Password rejected by server")
            } else {
                // Something weird happened
                throw AuthenticationException("Auth response ID does not match the original ID")
            }
        }
    }

    /**
     * Disconnect from the current server
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun disconnect() {
        synchronized(sync) { socket?.close() }
    }

    /**
     * Send a command to the server
     *
     * @param payload The command to send
     * @return The payload of the response
     *
     * @throws IOException
     */
    @Throws(Exception::class)
    fun command(payload: String): String {
        require(payload.trim { it <= ' ' }.isNotEmpty()) { "Payload can't be null or empty" }
        val response = send(RconPacket.SERVERDATA_EXECCOMMAND, payload.toByteArray())
        if (response != null) {
            return BotUtil.sendMessage(String(response.payload, charset)).contentToString()
        }
        return "发生异常"
    }

    @Throws(IOException::class)
    private fun send(type: Int, payload: ByteArray): RconPacket? {
        synchronized(sync) { return RconPacket.send(this, type, payload) }
    }

    init {
        // Default charset is utf8

        // Connect to host
        connect(host, port, password)
    }
}